Logic: generating truth table for boolean expression

konstantintogoi opened the pull request:
Added three functions: truth_table, term_to_integer, integer_to_term.

Function term_to_integer converts string or a list of ones and zeros to integer.

Function integer_to_term converts decimal to binary. Also this functions
can be used to obtain k-th term in n variables, and to obtain list of
truth values of k-th function of n variables.

Function truth_table returns a generator of all possible configurations of
the input variables, and the result of the boolean expression for those
values.

Example:

In [1]: from sympy.logic.boolalg import truth_table

In [2]: from sympy.abc import x, y

In [3]: table = truth_table(x ^ y, [x, y])

In [4]: for row in table:
   ...:     print 'term:', row[0], ', value:', row[1]
   ...:
term: [0, 0] , value: False
term: [0, 1] , value: True
term: [1, 0] , value: True
term: [1, 1] , value: False

konstantintogoi committed:
logic: generate truth table for boolean expression

Generating truth tables for boolean expressions.
Converting from binary to decimal.
Converting from decimal to binary.

Example:

In [1]: from sympy.logic.boolalg import term_to_integer

In [2]: term_to_integer('10011010')
Out[2]: 154

In [3]: from sympy.logic.boolalg import integer_to_term

In [4]: integer_to_term(777)
Out[4]: [1, 1, 0, 0, 0, 0, 1, 0, 0, 1]

In [5]: from sympy.logic.boolalg import truth_table

In [6]: truth_table('a | b | c', ['a', 'b', 'c'])
Out[6]: [0, 1, 1, 1, 1, 1, 1, 1]

moorepants commented:
Thanks for the contribution. Couple of comments:

1. These function do not seem to operate on SymPy objects. I'd expect `truth_table` to work on SymPy boolean expressions instead of strings.
2. You'll need to write some unit tests for these functions beyond the examples. See the files that start with `test_` for examples.

smichr commented on sympy/logic/boolalg.py:
```diff
@@ -1304,6 +1304,87 @@ def append_symbol(arg, symbols):
+	========
+	>>> from sympy.logic.boolalg import truth_table
+	>>> truth_table('a&b | b&c | c&a', ['a', 'b', 'c'])
+	[0, 0, 0, 1, 0, 1, 1, 1]
```
Maybe TableForm should be used for the output?

KoStiG replied:
Thanks for your feedback.

Truth table, i.e. complete list of all points in *B<sup>n</sup>* together with the value of a function *f* in each point is the most elementary way to define a Boolean function.

But there is a convenient, fast way to determine all truth value combinations for any number of variables. Called *lexicographical ordering*, it goes like this. Suppose there are *n* different variables. Then there will be *2<sup>n</sup>* different truth value combinations of these variables. We can replace each binary vector *(x<sub>0</sub>, x<sub>1</sub>, . . . , x<sub>n-1</sub>) in *B<sup>n</sup>* by its decimal representation *x = x<sub>0</sub>2<sup>n-1</sup> + x<sub>1</sub>2<sup>n-2</sup> + . . . + x<sub>n-1</sub>*.  So instead of writing *f(0, 0, . . . , 0)* we can write *f(0)*, instead of *f(0, 0, . . . , 1)* we can write *f(1)*, instead of *f(1, 1, . . . , 1)* we can write *f(2<sup>n-1</sup>)* and so on. This allows us to define a Boolean function as *f = [f(0), f(1), f(2), . . . , f(2<sup>n-1</sup>)]*. For instance, suppose we have a 3-variable Boolean function *f = [0, 0, 0, 1, 0, 1, 1, 1]*, then our truth table will be:

    >>> from sympy.logic.boolalg import truth_table
    >>> f = [0, 0, 0, 1, 0, 1, 1, 1]
    >>> table = zip(range(len(f)), f)
    >>> for t in table:
    ...    print(t)
    (0, 0)
    (1, 1)
    (2, 0)
    (3, 1)
    (4, 1)
    (5, 1)
    (6, 0)
    (7, 1)

It is equivalent to:

    >>> from sympy.logic.boolalg import integer_to_term
    >>> n = len(f)
    >>> n_bits = n.bit_length() - 1
    >>> table = zip(map(integer_to_term, range(n), [n_bits]*n), f)
    >>> for t in table:
    ...    print(t)
    ([0, 0, 0], 0)
    ([0, 0, 1], 1)
    ([0, 1, 0], 0)
    ([0, 1, 1], 1)
    ([1, 0, 0], 1)
    ([1, 0, 1], 1)
    ([1, 1, 0], 0)
    ([1, 1, 1], 1)

By using *lexicographical ordering* each truth value combination can be *efficiently encoded* as index of truth value.


Function <code>truth_table</code> do work on SymPy boolean expressions, not only strings:

    >>> from sympy.logic.boolalg import truth_table
    >>> from sympy.abc import a,b,c
    >>> expr = a ^ b >> c
    >>> expr
    Xor(a, Implies(b, c))
    >>> truth_table(expr, [a, b, c])
    [1, 1, 0, 1, 0, 0, 1, 0]

The variables must be given as the second argument in order to define which bit represents variable in truth value combinations. For example, <code>truth_table(a >> b, [b, a])</code> gives <code>[1, 0, 1, 1]</code> because variable <code>b</code> represents the first bit.


<code>integer_to_term</code> can be used to obtain the *k-th* term in *n* variables like this: <code>integer_to_term(k, n)</code>. For example, 13-th term in 5 variables:

    >>> from sympy.logic.boolalg import integer_to_term
    >>> integer_to_term(13, 5)
    [0, 1, 1, 0, 1]

Also, <code>integer_to_term</code> can be used to obtain truth values of the *k-th* function of *n* variables (There are *2<sup>2<sup>n</sup></sup>* functions of *n* variables). For example, you can easily print truth values of all 16 boolean functions of 2 variables like this:

    >>> from sympy.logic.boolalg import integer_to_term
    >>> n = 2
    >>> functions = map(integer_to_term, range(16), [4]*16)
    >>> for i in range(4):
    ...    s = ''
    ...    for j in range(16):
    ...        s = s + ' | ' + str(functions[j][i])
    ...    print(s)

    | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 1 | 1 | 1 | 1 | 1 | 1 | 1 | 1
    | 0 | 0 | 0 | 0 | 1 | 1 | 1 | 1 | 0 | 0 | 0 | 0 | 1 | 1 | 1 | 1
    | 0 | 0 | 1 | 1 | 0 | 0 | 1 | 1 | 0 | 0 | 1 | 1 | 0 | 0 | 1 | 1
    | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1

smichr replied:
> Also, integer_to_term can be used to obtain truth values

I followed you to here, but this example lost me. What is n used for? Is there a standard order for the 16 truth functions? Anyway, this isn't necessary (to me) since what preceded was clear.

KoStiG replied:
At first, I wanted explicitly show how number of input combinations and number of all possible functions depend on number of variables

    >>> from sympy.logic.boolalg import integer_to_term
    >>> n = 3
    >>> functions = map(integer_to_term, range(2*(2**n)), [2**n]*(2**(2**n))
    >>> for i in range(2**n):
    ...    s = ''
    ...    for j in range(2**(2**n)):
    ...        s = s + ' | ' + str(functions[j][i])
    ...    print(s)

smichr replied:
I must be missing something -- I don't work in this area often. By 'functions' do you mean things like And, Or, etc... I don't understand where the `2**(2**n)` is derived from. Can you point me to a reference, perhaps?

KoStiG replied:
Yes, by 'functions' I mean things like And, Or, etc.

Explanation of how all possible Boolean functions can be constructed on a set of Boolean variables - http://cs.fit.edu/~wds/classes/adm/Lectures/BooleanFunctions.pdf

smichr commented on sympy/logic/boolalg.py:
```diff
@@ -1304,6 +1304,87 @@ def append_symbol(arg, symbols):
+	[[0, 0, 0], [0, 1, 1], [1, 0, 0], [1, 0, 1], [1, 1, 1]]
+
+	"""
+	n = i.bit_length() if i else 1
```
bit_length is only available since Py 3.1 so we will have to perhaps define a bin function in core.compatibility and use that, e.g.

...
def bin(s):
    return str(s) if s<=1 else bin(s>>1) + str(s&1)
```

KoStiG replied:
There is built-in function [<code>bin()</code>](https://docs.python.org/release/2.6.1/library/functions.html#bin) available since Py 2.6.1. Can we use it?
If so, we can define function in core.compatibility, which is equivalent to <code>bit_length()</code>:

    >>> def bit_len(s):
    >>>     s = bin(s)
    >>>     s = s.lstrip('-0b')
    >>>     return len(s)

<code>str.lstrip()</code> is available since Py 2.2

smichr replied:
hmm...must have been looking at some older docs that recommended bin. I see you are right. Yes, we can use it. The oldest Py we support is 2.6.

I thought we generally try to make compatibility use functions that emulate a function in a newer version of sympy. What you are suggesting is making a function that emulates a method. I'm not sure this is a good idea because then we would have to deprecate the use of bit_len later. Perhaps it's best to just inline this as

```
# TODO: replace with `n = i.bit_length() if i else 1` when 2.X support is dropped
n = len(bin(i).lstrip('-0b')) if i else 1
```

smichr commented on sympy/logic/boolalg.py:
```diff
@@ -1304,6 +1304,87 @@ def append_symbol(arg, symbols):
+	"""
+	Gives a list of the base 2 digits in the integer i.
+	If additional n_bits argument is passed,
+	integer_to_term() extends list of digits with zeros.
```
extends ... -> prepends zeros to the list of digits.

smichr commented on sympy/logic/boolalg.py:
```diff
@@ -1304,6 +1304,87 @@ def append_symbol(arg, symbols):
+
+def truth_table(expr, variables):
+	"""
+	Gives a list of truth values of the boolean expression.
```
After reading your clarifications (which are much appreciated) I would suggest something like this:

```
Gives a list of truth values of the boolean expression in ``n`` variables where the i-th truth value is the output of the function when the input for each variable is ``integer_to_term(i, n)``.

EXAMPLES
========

The expression ``x >> y`` has 4 possible values corresponding to the
variables of x and y taking on values of 0 and 1:

>>> expr, vars = x >> y, [y, x]
>>> for i in range(2):
...     for j in range(2):
...         reps = list(zip(vars, (i, j)))
...         print reps, '->', int(bool(expr.subs(reps)))
...
[(y, 0), (x, 0)] -> 1
[(y, 0), (x, 1)] -> 0
[(y, 1), (x, 0)] -> 1
[(y, 1), (x, 1)] -> 1

This is represented more efficiently by the truth_table:

>>> truth_table(x >> y, [y, x])
[1, 0, 1, 1]

Note that the corresponding input values of variables can be deduced from the index of a
given output. For example, the False value at position 1 corresponds to an input
of y=0 and x=1:

>>> list(zip(vars, integer_to_term(1, len(vars))))
[(y, 0), (x, 1)]

The expression can also be given as a string,
...

```

KoStiG replied:
Thanks for your suggestions.

What do you think about using a generator as an output for <code>truth_table</code>?:

	def truth_table(expr, variables, input=True):
		"""
		Gives a generator of all possible configurations of the input variables,
		and the result of the boolean expression for those values.

		Parameters
		==========

		expr : string or boolean expression
		variables : list of variables
		input : boolean (default True)
			indicates whether to return the input combinations.

		Examples
		========
		>>> from sympy.logic.boolalg import truth_table
		>>> from sympy.abc import x,y
		>>> table = truth_table(x >> y, [x, y])
		>>> for t in table:
		>>>     print t[0], '->', t[1]
		[0, 0] -> True
		[0, 1] -> True
		[1, 0] -> False
		[1, 1] -> True

		>>> table = truth_table('x | y', ['x', 'y'])
		>>> print list(table)
		[([0, 0], False), ([0, 1], True), ([1, 0], True), ([1, 1], True)]

		If input is false, truth_table returns only a list of truth values.
		In this case, the corresponding input values of variables can be
		deduced from the index of a given output.

		>>> from sympy.logic.boolalg import integer_to_term
		>>> vars = [y, x]
		>>> values = truth_table(x >> y, vars, input=False)
		>>> values = list(values)
		>>> print values
		[True, False, True, True]

		>>> for i, value in enumerate(values):
		...     print zip(vars, integer_to_term(i, len(vars))), '->', value
		[(y, 0), (x, 0)] -> True
		[(y, 0), (x, 1)] -> False
		[(y, 1), (x, 0)] -> True
		[(y, 1), (x, 1)] -> True

		"""
		variables = [sympify(v) for v in variables]

		expr = sympify(expr)
		if not isinstance(expr, BooleanFunction):
			yield expr

		table = product([0, 1], repeat=len(variables))
		for term in table:
			term = list(term)
			value = expr.xreplace(dict(zip(variables, term)))

			if input:
				yield term, value
			else:
				yield value


Truth tables have the exponential growth in size as the number of variables increase. It would be better to iterate through truth table chunk by chunk, if the number of variables is large.

smichr commented:
I just had a few minor suggestions, @moorepants . Please merge when you think this is ready.

konstantintogoi committed:
logic: Generating truth table

Added three functions: truth_table, term_to_integer, integer_to_term.

Function term_to_integer converts string or a list of ones and zeros to integer.

Function integer_to_term converts decimal to binary. Also this functions
can be used to obtain k-th term in n variables, and to obtain list of
truth values of k-th function of n variables.

Function truth_table returns a generator of all possible configurations of
the input variables, and the result of the boolean expression for those
values.

Example:

In [1]: from sympy.logic.boolalg import truth_table

In [2]: from sympy.abc import x, y

In [3]: table = truth_table(x ^ y, [x, y])

In [4]: for row in table:
   ...:     print 'term:', row[0], ', value:', row[1]
   ...:
term: [0, 0] , value: False
term: [0, 1] , value: True
term: [1, 0] , value: True
term: [1, 1] , value: False

konstantintogoi committed:
Merge branch 'master' into truthtable

Conflicts:
	sympy/logic/tests/test_boolalg.py

konstantintogoi committed:
changing docstrings
